File: sdk/typescript/src/namespaces/pulse.ts
/**
 * Pulse Namespace API
 *
 * Provides access to trending topics from HackerNews, Reddit, Twitter, and other sources.
 * Use Pulse to discover timely debate topics and analyze sentiment across platforms.
 */

/**
 * Supported pulse sources
 */
export type PulseSource = 'hackernews' | 'reddit' | 'twitter' | 'all';

/**
 * Trending topic from pulse sources
 */
export interface TrendingTopic {
  id: string;
  title: string;
  url?: string;
  source: PulseSource;
  score: number;
  comments_count?: number;
  posted_at: string;
  fetched_at: string;
  category?: string;
  tags?: string[];
  sentiment?: 'positive' | 'negative' | 'neutral' | 'mixed';
}

/**
 * Debate suggestion based on trending topic
 */
export interface DebateSuggestion {
  topic: TrendingTopic;
  suggested_task: string;
  suggested_agents: string[];
  relevance_score: number;
  controversy_score: number;
  timeliness_score: number;
}

/**
 * Pulse analytics for debate outcomes on trending topics
 */
export interface PulseAnalytics {
  total_debates: number;
  consensus_rate: number;
  average_rounds: number;
  top_sources: Array<{ source: PulseSource; count: number; consensus_rate: number }>;
  trending_categories: Array<{ category: string; count: number }>;
  period: string;
}

/**
 * Pulse scheduler status
 */
export interface SchedulerStatus {
  running: boolean;
  paused: boolean;
  last_fetch?: string;
  next_fetch?: string;
  fetch_interval_minutes: number;
  topics_fetched: number;
  debates_started: number;
  errors_count: number;
}

/**
 * Scheduler configuration
 */
export interface SchedulerConfig {
  fetch_interval_minutes?: number;
  auto_debate?: boolean;
  min_score?: number;
  max_debates_per_hour?: number;
  sources?: PulseSource[];
  categories?: string[];
}

/**
 * Scheduled debate history entry
 */
export interface ScheduledDebate {
  id: string;
  topic_id: string;
  topic_title: string;
  source: PulseSource;
  debate_id?: string;
  started_at: string;
  completed_at?: string;
  result?: 'consensus' | 'no_consensus' | 'error';
}

/**
 * Interface for the internal client used by PulseAPI.
 */
interface PulseClientInterface {
  request<T>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Pulse API namespace.
 *
 * Provides methods for discovering and debating trending topics:
 * - Fetch trending topics from multiple sources
 * - Get debate suggestions based on current trends
 * - Manage the pulse scheduler for automated debates
 * - Analyze debate outcomes on trending topics
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai', apiKey: 'your-key' });
 *
 * // Get trending topics
 * const { topics } = await client.pulse.getTrending({ source: 'hackernews', limit: 10 });
 *
 * // Get a debate suggestion
 * const suggestion = await client.pulse.suggest();
 *
 * // Start a debate on a trending topic
 * const debate = await client.pulse.startDebate(suggestion.topic.id);
 * ```
 */
export class PulseAPI {
  constructor(private client: PulseClientInterface) {}

  /**
   * Get trending topics from pulse sources.
   */
  async getTrending(options?: {
    source?: PulseSource;
    category?: string;
    limit?: number;
    min_score?: number;
  }): Promise<{ topics: TrendingTopic[]; fetched_at: string }> {
    return this.client.request('GET', '/api/pulse/trending', { params: options });
  }

  /**
   * Get a debate suggestion based on current trending topics.
   */
  async suggest(options?: {
    source?: PulseSource;
    category?: string;
    prefer_controversial?: boolean;
  }): Promise<DebateSuggestion> {
    return this.client.request('GET', '/api/pulse/suggest', { params: options });
  }

  /**
   * Get analytics on debates started from trending topics.
   */
  async getAnalytics(options?: {
    source?: PulseSource;
    period?: '7d' | '30d' | '90d';
  }): Promise<PulseAnalytics> {
    return this.client.request('GET', '/api/pulse/analytics', { params: options });
  }

  /**
   * Start a debate on a trending topic.
   */
  async startDebate(
    topicId: string,
    options?: {
      agents?: string[];
      protocol?: Record<string, unknown>;
    }
  ): Promise<{ debate_id: string; task: string }> {
    return this.client.request('POST', '/api/pulse/debate-topic', {
      json: { topic_id: topicId, ...options },
    });
  }

  // =========================================================================
  // Scheduler Management
  // =========================================================================

  /**
   * Get the pulse scheduler status.
   */
  async getSchedulerStatus(): Promise<SchedulerStatus> {
    return this.client.request('GET', '/api/pulse/scheduler/status');
  }

  /**
   * Start the pulse scheduler.
   */
  async startScheduler(): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/pulse/scheduler/start');
  }

  /**
   * Stop the pulse scheduler.
   */
  async stopScheduler(): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/pulse/scheduler/stop');
  }

  /**
   * Pause the pulse scheduler (keeps state but stops fetching).
   */
  async pauseScheduler(): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/pulse/scheduler/pause');
  }

  /**
   * Resume a paused pulse scheduler.
   */
  async resumeScheduler(): Promise<{ success: boolean; message: string }> {
    return this.client.request('POST', '/api/pulse/scheduler/resume');
  }

  /**
   * Update the pulse scheduler configuration.
   */
  async updateSchedulerConfig(config: SchedulerConfig): Promise<{ success: boolean; config: SchedulerConfig }> {
    return this.client.request('PATCH', '/api/pulse/scheduler/config', {
      json: config as unknown as Record<string, unknown>,
    });
  }

  /**
   * Get the history of scheduled debates.
   */
  async getSchedulerHistory(options?: {
    platform?: PulseSource;
    limit?: number;
    offset?: number;
  }): Promise<{ debates: ScheduledDebate[]; total: number }> {
    return this.client.request('GET', '/api/pulse/scheduler/history', { params: options });
  }
}

File: sdk/python/aragora_sdk/namespaces/pulse.py
"""
Pulse Namespace API

Provides access to trending topics and real-time signals:
- Trending topics from HackerNews, Reddit, Twitter
- Debate topic suggestions based on signals
- Analytics on pulse data
- Scheduler management for automated pulse ingestion
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class PulseAPI:
    """
    Synchronous Pulse API for trending topics and signals.

    Pulse monitors external sources (HackerNews, Reddit, Twitter) for
    trending topics and suggests debate topics based on real-time signals.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> trending = client.pulse.get_trending(source="hackernews", limit=10)
        >>> topic = client.pulse.suggest_debate_topic()
        >>> scheduler = client.pulse.get_scheduler_status()
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # =========================================================================
    # Trending & Suggestions
    # =========================================================================

    def get_trending(
        self,
        source: Literal["all", "hackernews", "reddit", "twitter"] | None = None,
        category: str | None = None,
        limit: int = 20,
    ) -> dict[str, Any]:
        """
        Get trending topics from monitored sources.

        Args:
            source: Filter by source platform.
            category: Filter by category/domain.
            limit: Maximum number of topics to return.

        Returns:
            Dict with trending topics including scores, sources,
            and metadata.
        """
        params: dict[str, Any] = {"limit": limit}
        if source:
            params["source"] = source
        if category:
            params["category"] = category
        return self._client.request("GET", "/api/v1/pulse/trending", params=params)

    def suggest_debate_topic(
        self,
        source: str | None = None,
        category: str | None = None,
    ) -> dict[str, Any]:
        """
        Get a suggested debate topic from trending signals.

        Args:
            source: Prefer topics from this source.
            category: Prefer topics in this category.

        Returns:
            Dict with suggested topic, context, and reasoning.
        """
        params: dict[str, Any] = {}
        if source:
            params["source"] = source
        if category:
            params["category"] = category
        return self._client.request("GET", "/api/v1/pulse/suggest", params=params)

    def create_debate_topic(self, **kwargs: Any) -> dict[str, Any]:
        """
        Create a debate from a pulse topic.

        Args:
            **kwargs: Debate parameters including:
                - topic_id: Pulse topic to convert to debate
                - task: Override task description
                - agents: Specific agents to include

        Returns:
            Dict with created debate details.
        """
        return self._client.request("POST", "/api/v1/pulse/debate-topic", json=kwargs)

    # =========================================================================
    # Analytics
    # =========================================================================

    def get_analytics(self, **kwargs: Any) -> dict[str, Any]:
        """
        Get pulse analytics including topic distribution and trends.

        Returns:
            Dict with analytics data including topic counts by source,
            category distribution, and daily trends.
        """
        return self._client.request("POST", "/api/v1/pulse/analytics", json=kwargs)

    def get_scheduler_analytics(self) -> dict[str, Any]:
        """
        Get scheduler runtime metrics and store analytics.

        Returns combined scheduler metrics (polls, debates created/failed,
        uptime) and store analytics (by platform, by category, daily counts).

        Returns:
            Dict with scheduler metrics and store analytics.
        """
        return self._client.request("GET", "/api/v1/pulse/scheduler/analytics")

    # =========================================================================
    # Scheduler Management
    # =========================================================================

    def get_scheduler_status(self) -> dict[str, Any]:
        """
        Get pulse scheduler status.

        Returns:
            Dict with scheduler running state, next poll time,
            and configuration.
        """
        return self._client.request("POST", "/api/v1/pulse/scheduler/status")

    def get_scheduler_config(self) -> dict[str, Any]:
        """
        Get pulse scheduler configuration.

        Returns:
            Dict with scheduler configuration including poll interval,
            sources, and filtering rules.
        """
        return self._client.request("POST", "/api/v1/pulse/scheduler/config")

    def get_scheduler_history(self) -> dict[str, Any]:
        """
        Get pulse scheduler execution history.

        Returns:
            Dict with recent scheduler runs, their results, and timing.
        """
        return self._client.request("POST", "/api/v1/pulse/scheduler/history")

    def start_scheduler(self) -> dict[str, Any]:
        """
        Start the pulse scheduler.

        Returns:
            Dict with scheduler start confirmation.
        """
        return self._client.request("POST", "/api/v1/pulse/scheduler/start")

    def stop_scheduler(self) -> dict[str, Any]:
        """
        Stop the pulse scheduler.

        Returns:
            Dict with scheduler stop confirmation.
        """
        return self._client.request("POST", "/api/v1/pulse/scheduler/stop")

    def pause_scheduler(self) -> dict[str, Any]:
        """
        Pause the pulse scheduler.

        Returns:
            Dict with scheduler pause confirmation.
        """
        return self._client.request("POST", "/api/v1/pulse/scheduler/pause")

    def resume_scheduler(self) -> dict[str, Any]:
        """
        Resume a paused pulse scheduler.

        Returns:
            Dict with scheduler resume confirmation.
        """
        return self._client.request("POST", "/api/v1/pulse/scheduler/resume")


class AsyncPulseAPI:
    """
    Asynchronous Pulse API for trending topics and signals.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     trending = await client.pulse.get_trending(source="hackernews")
        ...     topic = await client.pulse.suggest_debate_topic()
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # =========================================================================
    # Trending & Suggestions
    # =========================================================================

    async def get_trending(
        self,
        source: Literal["all", "hackernews", "reddit", "twitter"] | None = None,
        category: str | None = None,
        limit: int = 20,
    ) -> dict[str, Any]:
        """Get trending topics from monitored sources."""
        params: dict[str, Any] = {"limit": limit}
        if source:
            params["source"] = source
        if category:
            params["category"] = category
        return await self._client.request("GET", "/api/v1/pulse/trending", params=params)

    async def suggest_debate_topic(
        self,
        source: str | None = None,
        category: str | None = None,
    ) -> dict[str, Any]:
        """Get a suggested debate topic from trending signals."""
        params: dict[str, Any] = {}
        if source:
            params["source"] = source
        if category:
            params["category"] = category
        return await self._client.request("GET", "/api/v1/pulse/suggest", params=params)

    async def create_debate_topic(self, **kwargs: Any) -> dict[str, Any]:
        """Create a debate from a pulse topic."""
        return await self._client.request("POST", "/api/v1/pulse/debate-topic", json=kwargs)

    # =========================================================================
    # Analytics
    # =========================================================================

    async def get_analytics(self, **kwargs: Any) -> dict[str, Any]:
        """Get pulse analytics including topic distribution and trends."""
        return await self._client.request("POST", "/api/v1/pulse/analytics", json=kwargs)

    async def get_scheduler_analytics(self) -> dict[str, Any]:
        """Get scheduler runtime metrics and store analytics."""
        return await self._client.request("GET", "/api/v1/pulse/scheduler/analytics")

    # =========================================================================
    # Scheduler Management
    # =========================================================================

    async def get_scheduler_status(self) -> dict[str, Any]:
        """Get pulse scheduler status."""
        return await self._client.request("POST", "/api/v1/pulse/scheduler/status")

    async def get_scheduler_config(self) -> dict[str, Any]:
        """Get pulse scheduler configuration."""
        return await self._client.request("POST", "/api/v1/pulse/scheduler/config")

    async def get_scheduler_history(self) -> dict[str, Any]:
        """Get pulse scheduler execution history."""
        return await self._client.request("POST", "/api/v1/pulse/scheduler/history")

    async def start_scheduler(self) -> dict[str, Any]:
        """Start the pulse scheduler."""
        return await self._client.request("POST", "/api/v1/pulse/scheduler/start")

    async def stop_scheduler(self) -> dict[str, Any]:
        """Stop the pulse scheduler."""
        return await self._client.request("POST", "/api/v1/pulse/scheduler/stop")

    async def pause_scheduler(self) -> dict[str, Any]:
        """Pause the pulse scheduler."""
        return await self._client.request("POST", "/api/v1/pulse/scheduler/pause")

    async def resume_scheduler(self) -> dict[str, Any]:
        """Resume a paused pulse scheduler."""
        return await self._client.request("POST", "/api/v1/pulse/scheduler/resume")
